### ISSUE TITLE
integrate lightning pose with anipose

### ISSUE THREAD

**Issue by siddypot** [danbider/lightning-pose#223]
I'm looking to convert the checkpoint file and .pkl to .h5 file format, has anyone come up with a solution for this?

**Comment by ksikka** [comment]
Hi @siddypot why do you need to do this?

**Comment by siddypot** [comment]
I would like to use [anipose](https://anipose.readthedocs.io/en/latest/) with lightning pose for 3d tracking. Anipose strictly accepts the .h5 file format

**Comment by ksikka** [comment]
Ok, I'll look into this.

**Comment by themattinthehatt** [comment]
@ksikka it might be worth reaching out to the anipose people instead and see if we can make a PR to allow anipose to accept the csv format (it really should anyways) - let's discuss tomorrow

**Comment by siddypot** [comment]
I am referring to the model weights. Lightning pose is generating the checkpoint file (.ckpt) and within the ckpt there is a .pkl file

**Comment by themattinthehatt** [comment]
@siddypot Maybe I am misunderstanding - anipose shouldn't need the LP checkpoint file. there is a larger anipose pipeline that runs inference with the pose estimation network and then runs triangulation on the pose estimation outputs. The first part will require LP integration with anipose, which we are currently thinking about, but it will be a bit more complex than just providing an LP checkpoint. On the other hand, you can run inference yourself with LP and then use the later part of the anipose pipeline to just run triangulation. That part will be easier to integrate. 

Can you describe your current workflow a bit more? Are you running inference on new videos with Lightning Pose yourself?

**Comment by siddypot** [comment]
After training LP I am left with 2d pose estimations in CSV. Anipose is expecting a DLC model. Based on the DLC model Anipose generates 2d pose estimation data in .csv, .pickle, and .h5 for all views and then based on that data anipose triangulates. Using LP I managed to get the .csv and the .pickle file, but anipose will not triangulate without all 3 files. I was hoping it would be a simple translation just to turn the LP data into a DLC model, or, as you said, take the 2d pose estimation data from LP and just use anipose for triangulation. Though no matter the step, I cannot get the LP data into Anipose, which is my biggest issue as of right now.

Sorry if I am not making too much sense, I am inexperienced with this technology.

**Comment by themattinthehatt** [comment]
No problem at all! We're very happy to make the integration between LP and Anipose much smoother. Can you point us to the place in the anipose code where you are running into issues?

**Comment by siddypot** [comment]
[model_folder in anipose config](https://anipose.readthedocs.io/en/latest/params.html) takes in the DLC project path. It would be great if we could get anipose to directly recognize the model folder of LP, but that may be a more difficult longer term project. 

[ `anipose analyze` (line 141) ](https://github.com/lambdaloop/anipose/blob/master/anipose/anipose.py) invokes [pose_videos](https://github.com/lambdaloop/anipose/blob/master/anipose/pose_videos.py) which is where I Cannot seem to figure out how to successfully implement LP

**Comment by themattinthehatt** [comment]
thanks for the pointers, we'll look into it, and get back to you early next week

**Comment by YitingChang** [comment]
I'm using anipose for triangulation and would like to follow this. 
I wrote a simple function to convert LP csv file to hdf file. 
`def lp2anipose(lp_path, anipose_path):
    df = pd.read_csv(lp_path, header = None, index_col = 0)
    # Convert object data to float data
    arr = df.iloc[3:].to_numpy()
    new_arr = arr.astype('f')
    new_df = pd.DataFrame(data=new_arr)
    # Create multi-level index for columns
    column_arr = df.iloc[0:3].to_numpy() 
    tuples = list(zip(*column_arr))
    new_df.columns = pd.MultiIndex.from_tuples(tuples, names=df.index[0:3])
    # Save in hdf format
    new_df.to_hdf(anipose_path, key = 'new_df', mode='w') `

**Comment by themattinthehatt** [comment]
thanks @YitingChang! I think you might be able to simplify this by doing
```
df = pd.read_csv(lp_path, header=[0, 1, 2], index_col=0)
df.to_hdf(anipose_path, key='new_df', mode='w')
```

**Comment by YitingChang** [comment]
Great! I will do that.

**Comment by siddypot** [comment]
> I'm using anipose for triangulation and would like to follow this. I wrote a simple function to convert LP csv file to hdf file.
> 
> ```
> def lp2anipose(lp_path, anipose_path):
>     df = pd.read_csv(lp_path, header = None, index_col = 0)
>     # Convert object data to float data
>     arr = df.iloc[3:].to_numpy()
>     new_arr = arr.astype('f')
>     new_df = pd.DataFrame(data=new_arr)
>     # Create multi-level index for columns
>     column_arr = df.iloc[0:3].to_numpy() 
>     tuples = list(zip(*column_arr))
>     new_df.columns = pd.MultiIndex.from_tuples(tuples, names=df.index[0:3])
>     # Save in hdf format
>     new_df.to_hdf(anipose_path, key = 'new_df', mode='w') 
> ```

@YitingChang Have you successfully triangulated using this h5 converter? If so could you provide documentation on how you did it? 

After getting the h5 files for my videos, running 

`## example triangulation without filtering, should take < 15 seconds
fname_dict = {
    'A': '2019-08-02-vid01-camA.h5',
    'B': '2019-08-02-vid01-camB.h5',
    'C': '2019-08-02-vid01-camC.h5',
}

d = load_pose2d_fnames(fname_dict, cam_names=cgroup.get_names())

score_threshold = 0.5

n_cams, n_points, n_joints, _ = d['points'].shape
points = d['points']
scores = d['scores']

bodyparts = d['bodyparts']

# remove points that are below threshold
points[scores < score_threshold] = np.nan

points_flat = points.reshape(n_cams, -1, 2)
scores_flat = scores.reshape(n_cams, -1)

p3ds_flat = cgroup.triangulate(points_flat, progress=True)
reprojerr_flat = cgroup.reprojection_error(p3ds_flat, points_flat, mean=True)

p3ds = p3ds_flat.reshape(n_points, n_joints, 3)
reprojerr = reprojerr_flat.reshape(n_points, n_joints)`

from [here](https://anipose.readthedocs.io/en/latest/aniposelib-tutorial.html#triangulation)

doesn't seem to do anything at all

**Comment by YitingChang** [comment]
@siddypot  Yes, I have successfully triangulated using this converter! I first create a configuration file. Then, I set the paths to data (see below) and use the triangulate function directly. 

config_file: path to the configuration file
calib_folder: path to the calibration folder
video_folder: path to the video folder
pose2d_folder: path to the 2d pose folder (h5 files)
output_fname: path to the output file (.csv)
```
from anipose.triangulate import triangulate
import toml

# Load config file
config = toml.load(config_file)

# Create file name dictionary 
pose_2d_files = glob(os.path.join(pose2d_folder, '*.h5'))
fname_dict = dict(zip(sorted(camera_names), sorted(pose_2d_files)))

# Triangulate
triangulate(config, calib_folder, video_folder, pose_folder,
                fname_dict, output_fname) 
```

**Comment by themattinthehatt** [comment]
@siddypot just wanted to check in to see if you've tried this out yet. I've talked with the anipose people and will work on integrating LP+Anipose sometime in January.

Btw would you mind telling me which lab you're from and what kind of data you're working with? The LP team is beginning to work a lot more functionality for multicamera setups so curious about the needs different people have.

**Comment by themattinthehatt** [comment]
@siddypot apologies for leaving this for so long - we've been hard at work on [Lightning Pose 3D](https://www.biorxiv.org/content/10.64898/2026.04.20.719731v1), which is now a preprint. As part of that we have created a brand new app that incorporates EKS, our own post-processing/triangulation code. I'm going to close this comment for now but there's still a possibility we could integrate with Anipose in the future.
